File: week10/23/prob1.py
# 375. Guess Number Higher or Lower II
class Solution(object):
    def getMoneyAmount(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp, length = [[0] * (n + 2) for _ in range(n + 2)], 2
        for i in range(1, n): dp[i][i + 1] = i
            
        while length < n:
            for i in range(1, n + 1 - length):
                dp[i][i + length] = min(
                    l + max(dp[i][l - 1], dp[l + 1][i + length]) for l in range(i, i + length + 1)
                )
                
            length += 1
            
        return dp[1][n]   

File: week10/23/prob1.cpp
// 375. Guess Number Higher or Lower II
class Solution {
public:
    int getMoneyAmount(int n) {
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, 0));

        function<int(int, int)> calculateCost = [&](int start, int end) {
            if (start >= end) return 0;
            if (dp[start][end] != 0) return dp[start][end];

            int minCost = INT_MAX;
            for (int guess = (start + end) / 2; guess <= end; ++guess) {
                int cost = guess + max(calculateCost(start, guess - 1), calculateCost(guess + 1, end));
                minCost = min(minCost, cost);
            }

            dp[start][end] = minCost;
            return minCost;
        };

        return calculateCost(1, n);
    }
};